Adding Device Simulator plugin for Input System support

RomanO-Unity opened the pull request:
This is a Device Simulator plugin which takes touchscreen inputs from the Device Simulator and inserts them into the Input System.

There are two issues that I am aware of:

1) Input System throws "A Native Collection has not been disposed, resulting in a memory leak. Enable Full StackTraces to get more details." This happens on domain reload sometimes. Enabling full stack trace does not improve the message.

I thought maybe there is some mismatch between touch begin and end/cancel events, but no everything matches and still the error persists. **Any ideas on this would be appreciated.**

2) IMGUI controls do not work, I've investigated why this happens and the fix needs to happen in the core not in this plugin.

RomanO-Unity committed:
Adding Device Simulator plugin for Input System support

RomanO-Unity committed:
Adding the meta file of the plugin

RomanO-Unity committed:
Adding UNITY_EDITOR ifdef for the plugin

RomanO-Unity commented:
@Rene-Damm I think the rest of the failures in tests are not caused by my change. What's the next step for this PR?

Rene-Damm commented:
> @Rene-Damm I think the rest of the failures in tests are not caused by my change.

Yup, only the configs marked "Required" need to be green.

> What's the next step for this PR?

I'll have a look at it.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Editor/DeviceSimulator/InputSytemPlugin.cs:
```diff
@@ -0,0 +1,92 @@
+using UnityEngine.InputSystem;
+using UnityEngine.InputSystem.LowLevel;
+
+public class InputSystemPlugin : DeviceSimulatorPlugin
```
Should be in namespace.

Also, can this be internal? If it has to be public, needs to be documented.

RomanO-Unity replied:
Yup, it can be internal, fixed it.

Rene-Damm reviewed (commented):
Couple comments.

- Ideally, this should have some level of automated testing in the package.
- Can we add docs to the manual for this?
- Needs a CHANGELOG entry.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Editor/DeviceSimulator/InputSytemPlugin.cs:
```diff
@@ -0,0 +1,92 @@
+        var playerSettings = Resources.FindObjectsOfTypeAll<PlayerSettings>()[0];
+        SerializedObject playerSettingsSerialized = new SerializedObject(playerSettings);
+
+        var activeInputHandler = playerSettingsSerialized.FindProperty("activeInputHandler").intValue;
```
Recommend using `EditorPlayerSettingsHelper`.

RomanO-Unity replied:
Done

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Editor/DeviceSimulator/InputSytemPlugin.cs:
```diff
@@ -0,0 +1,92 @@
+            m_DisabledDevices = new List<InputDevice>();
+            foreach (var device in InputSystem.devices)
+            {
+                if (device.native && device is Mouse && device.enabled)
```
What about `Pen` input?

RomanO-Unity replied:
Indeed it was misbehaving. Fixed it.

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Editor/DeviceSimulator/InputSytemPlugin.cs:
```diff
@@ -0,0 +1,92 @@
+
+            // UGUI elements like a button don't get pressed when multiple pointers for example mouse and touchscreen are sending data at the same time
+            m_DisabledDevices = new List<InputDevice>();
+            foreach (var device in InputSystem.devices)
```
This will not work correctly with devices that are added and removed on the fly.

RomanO-Unity replied:
Added code to handle it during InputSystem.onDeviceChange. Was not actually able to make it fire, all devices in the editor seem to be pre-added and disconnecting reconnecting does not seem to do anything...

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Editor/DeviceSimulator/InputSytemPlugin.cs:
```diff
@@ -0,0 +1,92 @@
+                InputSystem.RemoveDevice(m_SimulatorTouchscreen);
+            foreach (var device in m_DisabledDevices)
+            {
+                InputSystem.EnableDevice(device);
```
Device may have disappeared by now.

RomanO-Unity replied:
Added check to ignore devices that are not **added**.

RomanO-Unity committed:
Moved plugin into UnityEngine.InputSystem.Editor namespace and made it internal

RomanO-Unity committed:
Disabling Pen device, as it has similar conflict to a mouse. Using EditorPlayerSettingHelpers to know if Input System is enabled.

RomanO-Unity committed:
Disabling devices that are added on the fly

RomanO-Unity committed:
Changelog

RomanO-Unity committed:
Adding Device Simulator documentation

RomanO-Unity committed:
Adding tests for Device Simulator plugin

RomanO-Unity commented:
@Rene-Damm I've addressed the issues that you've raised, it's ready for a second look.

andrew-oc commented on Packages/com.unity.inputsystem/InputSystem/Editor/DeviceSimulator/InputSytemPlugin.cs:
```diff
@@ -0,0 +1,110 @@
+#if UNITY_EDITOR && UNITY_2021_1_OR_NEWER
```
Typo in file name: InputSy**s**temPlugin

RomanO-Unity replied:
Big Oof... Fixed now.

andrew-oc commented on Assets/Tests/InputSystem/Plugins/DeviceSimulatorTests.cs:
```diff
@@ -0,0 +1,83 @@
+        Assert.AreEqual(Touch.activeTouches.Count, 0);
+
+        plugin.OnDestroy();
+    }
```
To fix the native array not disposed error, you need to disable the EnhancedTouchSupport at the end of this test. This will dispose the native array that was allocated to hold touch state.

RomanO-Unity replied:
Done

andrew-oc commented on Assets/Tests/InputSystem/Plugins/DeviceSimulatorTests.cs:
```diff
@@ -0,0 +1,83 @@
+        var touch = new TouchEvent();
+        var type = typeof(TouchEvent);
+        object touchObject = touch;
+        var touchIdProperty = type.GetField("<touchId>k__BackingField", BindingFlags.Instance | BindingFlags.NonPublic);
```
Could we do something a little more robust here? Maybe use nameof() to get the property name and check that the backing fields actually exists?

RomanO-Unity replied:
Done what you suggested. Also added a few asserts, so hopefully if this code ever breaks due to changes to TouchEvent, errors will be clearer.

RomanO-Unity committed:
Fixed small issues raised by reviewers

- Fixed a typo in InputSystemPlugin file name
- Disabled EnhancedTouchSupport to clean it up properly
- Added more checks when reflecting upon a backing field of a property

andrew-oc committed:
Merge branch 'develop' into device-simulator/plugin

andrew-oc committed:
Try to get failing test to pass

andrew-oc committed:
Device simulator tests now implement InputTestFixture

This fixes broken tests when running from batch mode. Also swapped arguments in all Asset.AreEqual calls in device simulator tests because they were reversed.

andrew-oc committed:
Merge branch 'develop' into device-simulator/plugin
